From 360cbf10876fc964643418f6f1628fa75a48d6de Mon Sep 17 00:00:00 2001
From: Remi Herilier <remi.herilier@picviz.com>
Date: Fri, 11 Sep 2015 14:40:22 +0200
Subject: [PATCH 2/6] Permit to specify a default QTransform in QGraphisScene

As we use our own implementation of a QGraphicsView (PVGraphicsView), the
views transformation retrieval done by QGraphisScene does not work; the
default QTransform constructor is used instead, breaking the zoom.

We add new methods to set a default transformation to use as default one
---
 qtbase/src/widgets/graphicsview/qgraphicsscene.cpp | 25 ++++++++++++++++++++++++-
 qtbase/src/widgets/graphicsview/qgraphicsscene.h   |  3 +++
 qtbase/src/widgets/graphicsview/qgraphicsscene_p.h |  3 +++
 3 files changed, 30 insertions(+), 1 deletion(-)

diff --git a/qtbase/src/widgets/graphicsview/qgraphicsscene.cpp b/qtbase/src/widgets/graphicsview/qgraphicsscene.cpp
index bccdb1f..c3112f6 100644
--- a/qtbase/src/widgets/graphicsview/qgraphicsscene.cpp
+++ b/qtbase/src/widgets/graphicsview/qgraphicsscene.cpp
@@ -1117,7 +1117,7 @@ QList<QGraphicsItem *> QGraphicsScenePrivate::itemsAtPosition(const QPoint &scre
     Q_Q(const QGraphicsScene);
     QGraphicsView *view = widget ? qobject_cast<QGraphicsView *>(widget->parentWidget()) : 0;
     if (!view)
-        return q->items(scenePos, Qt::IntersectsItemShape, Qt::DescendingOrder, QTransform());
+        return q->items(scenePos, Qt::IntersectsItemShape, Qt::DescendingOrder, defaultViewTransform);
 
     const QRectF pointRect(QPointF(widget->mapFromGlobal(screenPos)), QSizeF(1, 1));
     if (!view->isTransformed())
@@ -5601,6 +5601,29 @@ void QGraphicsScene::setPalette(const QPalette &palette)
     d->setPalette_helper(resolvedPalette);
 }
 
+/*! \since 4.8.4 (Picviz)
+ *
+ * Set the default scene-to-view transformation used if none can be found.
+ * By default, an identity matrix is used.
+ */
+void QGraphicsScene::setDefaultViewTransform(QTransform const& viewTransform)
+{
+	Q_D(QGraphicsScene);
+	d->defaultViewTransform = viewTransform;
+}
+
+/*! \since 4.8.4 (Picviz)
+ *
+ * Get the default scene-to-view transformation used if none can be found.
+ *
+ * \sa setDefaultTransform
+ */
+QTransform const& QGraphicsScene::defaultViewTransform() const
+{
+	Q_D(const QGraphicsScene);
+	return d->defaultViewTransform;
+}
+
 /*!
     \since 4.6
 
diff --git a/qtbase/src/widgets/graphicsview/qgraphicsscene.h b/qtbase/src/widgets/graphicsview/qgraphicsscene.h
index cde0eda..3f39981 100644
--- a/qtbase/src/widgets/graphicsview/qgraphicsscene.h
+++ b/qtbase/src/widgets/graphicsview/qgraphicsscene.h
@@ -249,6 +249,9 @@ public:
 
     bool sendEvent(QGraphicsItem *item, QEvent *event);
 
+	void setDefaultViewTransform(QTransform const& viewTransform);
+	QTransform const& defaultViewTransform() const;
+
 public Q_SLOTS:
     void update(const QRectF &rect = QRectF());
     void invalidate(const QRectF &rect = QRectF(), SceneLayers layers = AllLayers);
diff --git a/qtbase/src/widgets/graphicsview/qgraphicsscene_p.h b/qtbase/src/widgets/graphicsview/qgraphicsscene_p.h
index 9e5bcec..b491b7d 100644
--- a/qtbase/src/widgets/graphicsview/qgraphicsscene_p.h
+++ b/qtbase/src/widgets/graphicsview/qgraphicsscene_p.h
@@ -69,6 +69,7 @@
 #include <QtCore/qset.h>
 #include <QtGui/qfont.h>
 #include <QtGui/qpalette.h>
+#include <QtGui/qtransform.h>
 #include <QtWidgets/qstyle.h>
 #include <QtWidgets/qstyleoption.h>
 
@@ -181,6 +182,8 @@ public:
     QList<QGraphicsItem *> cachedItemsUnderMouse;
     QList<QGraphicsItem *> hoverItems;
     QPointF lastSceneMousePos;
+
+	QTransform defaultViewTransform;
     void enableMouseTrackingOnViews();
     QMap<Qt::MouseButton, QPointF> mouseGrabberButtonDownPos;
     QMap<Qt::MouseButton, QPointF> mouseGrabberButtonDownScenePos;
-- 
2.5.1

